src/pudl/extract/ferc1.py

```diff
@@ -213,429 +208,272 @@
-    truncated column names extracted from that DBF file, and the full column names
-    extracted from the DBC file. Read the column datatypes out of each DBF file and use
-    them to define the columns in the new Table object.
 
-    Args:
-        table_name: The name of the new table to be added to the database schema.
-        sqlite_meta: The database schema to which the newly defined
-            :class:`sqlalchemy.Table` will be added.
-        dbc_map: A dictionary of dictionaries
-        ferc1_dbf_ds: Initialized FERC1 DBF datastore.
-        refyear: Reference year to use as a template for the database schema.
-    """
-    if refyear is None:
-        refyear = max(DataSource.from_id("ferc1").working_partitions["years"])
+    def finalize_schema(self, meta: sa.MetaData) -> sa.MetaData:
+        """This method is called just before the schema is written to sqlite.
 
-    new_table = sa.Table(table_name, sqlite_meta)
+        You can use this method to apply dataset specific alterations to the schema,
+        such as adding primary and foreign key constraints.
+        """
+        return meta
 
-    dbf_filename = DBF_TABLES_FILENAMES[table_name]
-    filedata = ferc1_dbf_ds.get_file(refyear, dbf_filename)
+    def postprocess(self):
+        """This metod is called after all the data is loaded into sqlite."""
+        pass
 
-    ferc1_dbf = DBF(dbf_filename, ignore_missing_memofile=True, filedata=filedata)
 
-    # Add Columns to the table
-    for field in ferc1_dbf.fields:
-        if field.name == "_NullFlags":
-            continue
-        col_name = dbc_map[table_name][field.name]
-        col_type = DBF_TYPES[field.type]
-        if col_type == sa.String:
-            col_type = sa.String(length=field.length)
-        new_table.append_column(sa.Column(col_name, col_type))
+class Ferc1FoxProExtractor(FoxProExtractor):
+    """Wrapper for running the foxpro to sqlite conversion of FERC1 dataset."""
 
-    col_names = [c.name for c in new_table.columns]
+    DATABASE_NAME = "ferc1.sqlite"
 
-    if table_name == "f1_respondent_id":
-        new_table.append_constraint(
-            sa.PrimaryKeyConstraint("respondent_id", sqlite_on_conflict="REPLACE")
-        )
+    def get_datastore(self, base_datastore: Datastore) -> FercFoxProDatastore:
+        """Returns the instace of FoxProDatastore.
 
-    if ("respondent_id" in col_names) and (table_name != "f1_respondent_id"):
-        new_table.append_constraint(
-            sa.ForeignKeyConstraint(
-                columns=[
-                    "respondent_id",
-                ],
-                refcolumns=["f1_respondent_id.respondent_id"],
-            )
+        This wraps the generic base_datastore and constructs instance of
+        FercFoxProDatastore.
+        """
+        # TODO(rousik): mapping from dataset name to dbc_filename may be moved to constants or kept somerwhere with the rest
```

Comment:
It seems like `F1_PUB.DBC` and its equivalents for the other data sources should definitely be in the `package_data` that maps out the files & directories, as we've done with the Excel spreadsheets. If the root directory that we need to look for all the DBF files in is always defined by wherever the DBC file is, then maybe we just store the path to the DBC path, and drop the filename to get the directory path (which I think is being stored independently now).

Reply:
🤔 yeah, this sounds like the good way to go about that.

Reply:
I think this is now done. Marking as resolved.